examples/cis3-nft-sponsored-txs/src/lib.rs

```diff
@@ -0,0 +1,2076 @@
+        }
+    }
+}
+
+/// The contract state.
+// Note: The specification does not specify how to structure the contract state
+// and this could be structured in a more space efficient way depending on the use case.
+#[derive(Serial, DeserialWithState, StateClone)]
+#[concordium(state_parameter = "S")]
+struct State<S> {
+    /// The state for each address.
+    state:               StateMap<Address, AddressState<S>, S>,
+    /// All of the token IDs
+    all_tokens:          StateSet<ContractTokenId, S>,
+    /// Map with contract addresses providing implementations of additional
+    /// standards.
+    implementors:        StateMap<StandardIdentifierOwned, Vec<ContractAddress>, S>,
+    /// A registry to link an account to an public key and its nonce. The
+    /// corresponding private key registered here has full access to the
+    /// tokens controlled by the account. The nonce is used to prevent replay
+    /// attacks of signed transactions.
+    public_key_registry: StateMap<AccountAddress, (PublicKeyEd25519, u64), S>,
+}
+
+/// Part of the parameter type for the contract function `registerPublicKey`.
+/// Takes the account and the public key that should be linked.
+#[derive(Debug, Serialize, SchemaType)]
+pub struct RegisterPublicKeyParam {
+    /// Account that a public key will be registered to.
+    account:    AccountAddress,
+    /// The public key that should be linked to the above account.
+    public_key: PublicKeyEd25519,
+}
+
+/// The parameter type for the contract function `setImplementors`.
+/// Takes a standard identifier and list of contract addresses providing
+/// implementations of this standard.
+#[derive(Debug, Serialize, SchemaType)]
+struct SetImplementorsParams {
+    /// The identifier for the standard.
+    id:           StandardIdentifierOwned,
+    /// The addresses of the implementors of the standard.
+    implementors: Vec<ContractAddress>,
+}
+
+/// Part of the parameter type for the contract function `permit`.
+/// Specifies the transaction payload that should be forwarded to either the
+/// `transfer` or the `updateOperator` function.
+#[derive(Debug, Serialize, SchemaType, Clone)]
+enum PermitPayload {
+    Transfer(TransferParameter),
+    UpdateOperator(UpdateOperatorParams),
+}
+
+/// Part of the parameter type for the contract function `permit`.
+/// Specifies the message that is signed.
+#[derive(SchemaType, Serialize)]
+struct PermitMessage {
+    /// The contract_address that the signature is intended for.
+    contract_address: ContractAddress,
+    /// The entry_point that the signature is intended for.
+    entry_point:      OwnedEntrypointName,
+    /// A nonce to prevent replay attacks.
+    nonce:            u64,
+    /// A timestamp to make signatures expire.
+    timestamp:        Timestamp,
+    /// The payload that should be forwarded to either the `transfer` or the
+    /// `updateOperator` function.
+    payload:          PermitPayload,
+}
```

Comment:
So I think there is a bit of an annoying issue here.

The serialization of `PermitPayload` will have a one byte tag, and then the transfer params or update operator params.

This means that in the CIS3 standard we would have to specify that the "payload" should have this extra byte tag or something like it instead of just the payload of the entrypoint. That's not necessarily bad, but needs to be clarified.

What exactly do we want here? Do we want to allow the permit to add extra data in addition to the parameters of the entrypoints?

Reply:
Could we make the payload generic and leave it to the implementation?

Reply:
If we allow any type of payload, then we can completely remove the link to CIS-2, right?
Then CIS-3 can become a universal standard for sponsored transactions, which I think is useful.